src/bit_encoding/bititer.rs

```diff
@@ -91,45 +110,39 @@ impl<I: Iterator<Item = u8>> BitIter<I> {
         Self::from(iter)
     }
 
-    /// Reads up to 64 bits as a big-endian number
-    ///
-    /// `n` must be between 0 and 63 inclusive; the function will panic
-    /// otherwise.
-    ///
-    /// Returns `None` if the underlying byte iterator runs out
-    pub fn read_bits_be(&mut self, mut n: usize) -> Option<u64> {
-        assert!(n < 64);
-        if n == 0 {
-            return Some(0);
+    /// Reads a single bit from the iterator.
+    pub fn read_bit(&mut self) -> Result<bool, EarlyEndOfStreamError> {
+        self.next().ok_or(EarlyEndOfStreamError)
+    }
+
+    /// Reads two bits from the iterator.
+    pub fn read_u2(&mut self) -> Result<u2, EarlyEndOfStreamError> {
+        match (self.next(), self.next()) {
+            (Some(false), Some(false)) => Ok(u2::_0),
+            (Some(false), Some(true)) => Ok(u2::_1),
+            (Some(true), Some(false)) => Ok(u2::_2),
+            (Some(true), Some(true)) => Ok(u2::_3),
+            _ => Err(EarlyEndOfStreamError),
         }
+    }
 
-        // 0 <= avail_bits <= 8
-        // 0 <= n < 64
-        let avail_bits = 8 - self.read_bits;
-        if avail_bits < n {
-            n -= avail_bits;
-            let mask = if avail_bits < 8 {
-                (1 << avail_bits) - 1
-            } else {
-                0xff
-            };
-            let pre_result = ((self.cached_byte & mask) as u64) << n;
-            self.cached_byte = self.iter.next()?;
-            self.read_bits = 0;
+    /// Reads a byte from the iterator.
+    pub fn read_u8(&mut self) -> Result<u8, EarlyEndOfStreamError> {
+        debug_assert!(self.read_bits > 0);
+        let cached = self.cached_byte;
+        self.cached_byte = self.iter.next().ok_or(EarlyEndOfStreamError)?;
 
-            if let Some(read) = self.read_bits_be(n) {
-                self.total_read += avail_bits;
-                Some(pre_result + read)
-            } else {
-                None
-            }
-        } else {
-            // 0 <= n <= 8
-            self.read_bits += n;
-            self.total_read += n;
-            let mask = if n < 8 { (1 << n) - 1 } else { 0xff };
-            Some(((self.cached_byte >> (avail_bits - n)) & mask) as u64)
+        Ok(cached.checked_shl(self.read_bits as u32).unwrap_or(0)
```

Comment:
54e6cf2bc70669ef6dda71647e2db015339d2f59: Are we updating `self.read_bits` anywhere?

Reply:
Not in this function, no. `read_bits` is how much of the current byte we've read ... and if we read an entire byte, this number doesn't change.

Reply:
Ah, that makes a lot of sense. I mixed it up with `total_read`.